### PR TITLE
Improve `HazelcastVersionLocator` (test utility)

### PR DESCRIPTION
**Pull request by JackPGreen**
`HazelcastVersionLocator` is used in the tests to download alternative versions of Hazelcast via Maven.

When I ran the tests, this failed, leading me to refactor for the following improvements:

- Support paths with spaces in the name (i.e. don't call `.split(" ")` before passing the command to `ProcessBuilder`)
- Look outside of the `$PATH` for Maven installations on OSX
- Dynamic resolution of the Maven repository rather than hardcoding to `~\.m2\repository`
- Resolve Maven artifact paths within the repository using Maven's `maven-resolver-provider` rather than manual path substitution
- Migrate static methods into the `Artifact` enum and make object-oriented
- Remove warning on artifact download that "speed up the {...} tests by installing the missing artifacts {...} so they don't have to be downloaded next time" - the command it was saying to run was the one it'd just run, so it's pointless

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/6343

### CODE REVIEW

**Review by JamesHazelcast — COMMENTED** [review]
Nice catch and solution Jack - I'm not a big fan of the `Process` usage, but as your comments show, the other "proper" solutions have created their own issues. Hopefully you're able to get some SO wisdom on resolving that to clean things up a bit 👍

File: hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java:56]: I've fixed the typo, but I've left the space formatting alone as that's how the IDE auto-formatted it.

File: hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java:57]: > Do we want to update this for 5.0? Makes sense imo

I don't see any value in changing it to v5? This test is testing a test utility, that it can download an arbitrary version of the required dependency successfully.

When it's used it's typically testing that older (i.e. v3 or v4) instances can be communicated with so it's fine I think.

File: hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java:44]: > This looks like an auto-format from your IDE - generally we avoid formatting changes (if the checkstyle passes, we're happy), so that way it does not cloud the PR's changes. Also basically all existing `@Category` annotations do not contain spaces around the curly braces.
> 
> If you haven't already, I suggest [setting up your IDE to use the Hazelcast checkstyle](https://hazelcast.atlassian.net/wiki/spaces/EN/pages/5374373/Codestyle+Code-Formatter) for local formatting.

I'm not using IntelliJ so I'm not able to set this up, but we have a mix of formatting across the codebase with regards to array initialisers having or not having spaces - the '@SuppressWarnings' are done in both ways.
I've updated my style to suit.

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:101]: > This warning about artifact downloads has been removed - was this intentional? Seems like a useful warning for local usage.

I removed it because I don't understand why it's there (risk of invoking Chesterton's Fence here).

The warning appears just before it runs a `mvn dependency:get`. That will download the version, if not already present, and store it for next time.
So telling you "next time, run the `mvn dependency:get` command it won't redownload it" is just wrong - because it's just done that.
If the Maven command downloaded to some temporary directory it'd make sense, or if it was a `SNAPSHOT` dependency (and then the test command didn't force updates). But it's not. So I don't understand.

Reminds me though, I need to update the EE readme as well.

File: hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java
Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java]: The extra space indents it under the `TODO` so it's all formatted as a TODO (which gives it all the TODO highlighting in the IDE) - without the extra space, the top line is highlighted to me while the 2nd is not.

File: hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java
Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java]: Just makes sense to keep it updated with our latest major version, since older major version will be forgotten about eventually - not a big deal at all though, happy to approve without as like you said the focus is on downloading _an artefact_ more than specifics.

**Review by JamesHazelcast — APPROVED** [review]
Thanks for the minor changes Jack - I'm still not a fan of the Process usage, it's probably a good idea to create a backlog JIRA to re-visit this and clean it up in the future.

**Comment by JackPGreen** [comment]
> Thanks for the minor changes Jack - I'm still not a fan of the Process usage, it's probably a good idea to create a backlog JIRA to re-visit this and clean it up in the future.

@vbekiaris has been working up some magic to get it working.

File: hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/starter/test/HazelcastVersionLocatorTest.java]: > The extra space indents it under the `TODO` so it's all formatted as a TODO (which gives it all the TODO highlighting in the IDE) - without the extra space, the top line is highlighted to me while the 2nd is not.

I think thats an IntelliJ-specific feature.
I did look in the rest of the codebase and the first multi-line `TODO` I found wasn't formatted like that (`com.hazelcast.cache.impl.operation.CacheEntryProcessorOperation.getBackupOperation()`). Which is handy as there's no way I could get my IDE to format like that automatically.

**Comment by vbekiaris** [comment]
Since this has been stalling for some time and I don't see an easy way forward, here is an update from my side:

- this PR relies on executing `mvn` via `ProcessBuilder` API. I am not a fan of spawning processes. More importantly, the `mvn` command location logic is insufficient because it attempts to locate `mvn`/`mvn.cmd` in the current working directory which is most probably wrong (cwd will be probably the "hazelcast" repository from which tests are executed). The fallback for macOS works for those who installed mvn via homebrew but eg fails for me (I use sdkman) and I suspect the same will happen with people on linux. Maybe, for unix-like systems, we could start a shell and execute `` `which mvn` ``. But even then. there are too many moving parts that may fail and I think we'll get into more trouble than benefit.
- we tried to get rid of the process spawning parts by directly using the maven resolver API and artifact resolution worked well [in this commit](https://github.com/vbekiaris/hazelcast/commit/e3f107ee813897a81975be0ec2a941618a60e24a). However, our OSGi testing dependencies include a 7-year-old `org.ops4j.pax.runner:pax-runner-no-jcl` artifact that shades (without relocating to another package) very old versions of aether and maven classes, breaking the test suite.
- there is no obvious way to upgrade our pax-exam OSGi tests so they can be executed and not mess with the classpath.

My suggestion is to not merge this PR as is. Let's create an issue to upgrade our OSGI deps and once we figure that out, merge the solution that uses maven resolver API and does not depend on process/environment. wdyt?

**Comment by JackPGreen** [comment]
> More importantly, the `mvn` command location logic is insufficient because it attempts to locate `mvn`/`mvn.cmd` in the current working directory which is most probably wrong (cwd will be probably the "hazelcast" repository from which tests are executed). The fallback for macOS works for those who installed mvn via homebrew but eg fails for me (I use sdkman) and I suspect the same will happen with people on linux.

I added the Homebrew fallback logic because [Eclipse wasn't looking up things on the path](https://stackoverflow.com/q/76866453), but as you've pointed out it doesn't cover every case.

I've removed this so that that bit works as before which prevents these issues for you.

3a5a11d2a7985d5c0204a8c7e649b1cf04007e82

> My suggestion is to not merge this PR as is

Does this change based on the above? I think it's still an improvement, albeit in an area that still needs more work.

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:125]: add `builder.add("--quiet");` (and probably `--batch-mode` too) here, otherwise the forked JVM runner complains because the maven command output in stdout messes with surefire's std[in,out] streams-based inter-process communication (AFAIK):

```
[WARNING] Corrupted channel by directly writing to native stream in forked JVM 1. See FAQ web page and the dump file .../hazelcast/hazelcast/target/surefire-reports/2023-09-13T15-31-34_745-jvmRun1.dumpstream
```

**Review by vbekiaris — APPROVED** [review]
please address the small comment I added and then we can merge, but let's keep an eye on builds for the next few days

File: hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastVersionLocator.java:125]: 9c9b2fa6c115aeed9368910d03cc6b464eeebe71

**Pull request merged**
